src/MatrixFields/field_name_dict.jl

```diff
@@ -237,6 +277,96 @@ function Base.one(matrix::FieldMatrix)
     return FieldNameDict(inferred_diagonal_keys, entries)
 end
 
+"""
+    get_field_first_index_offset(name::FieldName, ::Type{T}, ::Type{S})
+
+Returns the offset of the the field with name `name` in an object of type `S`
+in multiples of `sizeof(T)`.
+"""
+function get_field_first_index_offset(
+    name::FieldName,
+    ::Type{T},
+    ::Type{S},
+) where {T, S}
+    if name == @name()
+        return 0
+    end
+    child_name = extract_first(name)
+    child_type = fieldtype(S, child_name)
+    remaining_field_chain = drop_first(name)
+    field_index =
+        unrolled_filter(i -> fieldname(S, i) == child_name, 1:fieldcount(S))[1]
+    return DataLayouts.fieldtypeoffset(T, S, field_index) +
+           get_field_first_index_offset(remaining_field_chain, T, child_type)
+end
+if hasfield(Method, :recursion_relation)
+    dont_limit = (args...) -> true
+    for m in methods(get_field_first_index_offset)
+        m.recursion_relation = dont_limit
+    end
+end
+
+"""
+    get_scalar_keys(dict::FieldMatrix)
+
+Returns a `FieldMatrixKeys` object that contains the keys of all the scalar
+entries in the `FieldMatrix` `dict`.
+"""
+function get_scalar_keys(dict::FieldMatrix)
+    keys_tuple = unrolled_flatmap(keys(dict).values) do key
+        _, entry = unrolled_filter(pair -> key == pair[1], pairs(dict))[1]
+        entry =
+            entry isa ColumnwiseBandMatrixField ? entry.entries.:(1) : entry
+        unrolled_map(
+            filtered_child_names(
+                field -> eltype(field) <: Number,
+                entry,
+                @name()
+            ),
+        ) do name
+            (append_internal_name(key[1], name), key[2])
+        end
+    end
+    return FieldMatrixKeys(keys_tuple)
+end
+
+"""
+    scalar_fieldmatrix(field_matrix::FieldMatrix)
+
+Constructs a `FieldNameDict` where the keys and entries are views
+of the entries of `field_matrix`, which corresponding to the
+scalar components of entries of `field_matrix`.
+
+# Example usage
```

Comment:
I love the names in this example, but maybe it would be a bit clearer if it had names that are recognizable to users of ClimaAtmos, like `ρ` and `uₕ`? And maybe `foo` could be called something like `atmos_state`? This version is quite funny though so use whichever names you prefer.